lib/services/plugins.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"bytes"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// MarshalPlugin marshals Plugin resource to JSON.
+func MarshalPlugin(plugin types.Plugin, opts ...MarshalOption) ([]byte, error) {
+	if err := plugin.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	switch plugin := plugin.(type) {
+	case *types.PluginV1:
+		if !cfg.PreserveResourceID {
+			copy := *plugin
+			copy.SetResourceID(0)
+			plugin = &copy
+		}
+		var buf bytes.Buffer
+		err := (&jsonpb.Marshaler{}).Marshal(&buf, plugin)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return buf.Bytes(), nil
+	default:
+		return nil, trace.BadParameter("unsupported plugin resource %T", plugin)
+	}
+}
+
+// UnmarshalPlugin unmarshals the plugin resource from JSON.
+func UnmarshalPlugin(data []byte, opts ...MarshalOption) (types.Plugin, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("missing plugin resource data")
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var h types.ResourceHeader
+	if err := utils.FastUnmarshal(data, &h); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	switch h.Version {
+	case types.V1:
+		var plugin types.PluginV1
+		if err := jsonpb.Unmarshal(bytes.NewReader(data), &plugin); err != nil {
```

Comment:
Either the field is `Version` (as expected by `jsonpb.Unmarshal`) or it's `version` (as expected by `encoding/json`).